cpp/src_prims/distance/euclidean.cuh

```diff
@@ -55,13 +208,190 @@ void euclideanAlgo1(Index_ m, Index_ n, Index_ k, const InType *pA,
+      GemmConfig_, EpilogueFunctor_, Index_, EpilogueTraitsHelper_>
+      GemmEpilogueTraits_;
+    typedef ExpandedDistanceGemmEpilogue<GemmEpilogueTraits_> GemmEpilogue_;
+    typedef typename EpilogueFunctor_::Params EpiParams;
+
+    cublasOperation_t transa, transb;
+    const InType *aPtr, *bPtr;
+    Index_ lda, ldb, ldd;
+    Index_ gemm_m, gemm_n;
+    InType *rvec, *cvec;
+
+    transa = CUBLAS_OP_N;
+    transb = CUBLAS_OP_T;
+    aPtr = pA;
+    bPtr = pB;
+    lda = m;
+    ldb = n;
+    ldd = m;
+    gemm_m = m;
+    gemm_n = n;
+    cvec = row_vec;
+    rvec = col_vec;
+
+    LinAlg::gemm<InType, AccType, EffOutType, OutputTile_,
+                 AccumulatorsPerThread_, MainLoopFunctor_, Index_, GemmConfig_,
+                 EpilogueFunctor_, GemmEpilogueTraits_, GemmEpilogue_>(
+      transa, transb, gemm_m, gemm_n, k, (EffOutType)1, aPtr, lda, bPtr, ldb,
+      (EffOutType)0, nullptr, ldd, pDCast,
+      [cvec, rvec, enable_sqrt] HD(EpiParams & p) {
+        int err = p.initializeExtra(cvec, rvec, enable_sqrt);
+        return err;
+      },
+      fin_op, stream);
+  }
+}
+
+/**
+ * @brief the unexpanded euclidean distance matrix calculation kernel
+ *  It computes the following equation: cij = op((ai-bj)^2)
+ * @tparam DataT          input data-type (for A and B matrices)
+ * @tparam AccT           accumulation data-type
+ * @tparam OutT           output data-type (for C and D matrices)
+ * @tparam IdxT           index data-type
+ * @tparam sqrt           if the square root is computed or not
+ * @tparam Policy         struct which tunes the Contraction kernel
+ * @tparam CoreLambda     lambda which implements accumulation operation
+ * @tparam EpilogueLambda lambda which implements operation for calculating
+                          final value.
+ * @tparam FinalLambda    final lambda called on final distance value
+ *
+ * @param[in]       x input matrix
+ * @param[in]       y input matrix
+ * @param[in]       m number of rows of A and C/D
+ * @param[in]       n number of columns of B and C/D
+ * @param[in]       k number of cols of A and rows of B
+ * @param[output]   pD output matrix
+ * @param core_op   the core lambda
+ * @param epilog_op the epilogue lambda
+ * @param fin_op    the final gemm epilogue lambda
+ */
+template <typename DataT, typename AccT, typename OutT, typename IdxT,
+          bool Sqrt, typename Policy, typename CoreLambda,
+          typename EpilogueLambda, typename FinalLambda>
+__global__ __launch_bounds__(Policy::Nthreads, 2) void euclideanUnExpKernel(
+  const DataT *x, const DataT *y, IdxT m, IdxT n, IdxT k, OutT *dOutput,
+  CoreLambda core_op, EpilogueLambda epilog_op, FinalLambda fin_op) {
+  extern __shared__ char smem[];
+
+  PairwiseDistances<DataT, AccT, OutT, IdxT, Sqrt, Policy, CoreLambda,
+                    EpilogueLambda, FinalLambda>
+    obj(x, y, m, n, k, dOutput, smem, core_op, epilog_op, fin_op);
+  obj.run();
+}
```

Comment:
these kernels seem to be pretty much the same across multiple metrics. How about moving this too inside pairwise distances base header file?

Reply:
makes sense!